spec/index.html

```diff
@@ -3531,30 +3531,35 @@ <h2>A Canonical form of N-Quads</h2>
-    <li>Within <a data-cite="N-QUADS#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>,
-      the characters
-      <code>U+0008</code> (<code title="BACKSPACE"><sub>BS</sub></code>),
-      <code>U+0009</code> (<code title="HORIZONTAL TAB"><sub>HT</sub></code>),
-      <code>U+000A</code> (<code title="LINE FEED"><sub>LF</sub></code>),
-      <code>U+000C</code> (<code title="FORM FEED"><sub>FF</sub></code>),
-      <code>U+000D</code>  (<code title="CARRIAGE RETURN"><sub>CR</sub></code>),
-      <code>U+0022</code> (<code title="DOUBLE QUOTE">&quot;</code>), and
-      <code>U+005C</code> (<code title="BACKSLASH">\</code>)
-      MUST be encoded using <code><a data-cite="N-QUADS#grammar-production-ECHAR">ECHAR</a></code>.
-      Characters in the range from <code>U+0000</code> to <code>U+001F</code>
-      and <code>U+007F</code>  (<code title="delete"><sub>DEL</sub></code>)
-      that are not represented using <code><a data-cite="N-QUADS#grammar-production-ECHAR">ECHAR</a></code>
-      MUST be represented by <code><a data-cite="N-QUADS#grammar-production-UCHAR">UCHAR</a></code>.
-      All other characters MUST be represented by their native [[UNICODE]] representation.</li>
-    <li>The token <code><a data-cite="N-QUADS#grammar-production-EOL">EOL</a></code> MUST be a single <code>U+000A</code>.</li>
-    <li>The final <code><a data-cite="N-QUADS#grammar-production-EOL">EOL</a></code> MUST be provided.</li>
+    <li>Within <a data-cite="N-QUADS#grammar-production-STRING_LITERAL_QUOTE">STRING_LITERAL_QUOTE</a>:
+      <ul>
+        <li>Characters
+          <code>BS</code> (backspace, code point <code class="codepoint">U+0008</code>),
+          <code>HT</code> (horizontal tab, code point <code class="codepoint">U+0009</code>),
+          <code>LF</code> (line feed, code point <code class="codepoint">U+000A</code>),
+          <code>FF</code> (form feed, code point <code class="codepoint">U+000C</code>),
+          <code>CR</code> (carriage return, code point <code class="codepoint">U+000D</code>),
+          <code>&quot;</code> (quotation mark, code point <code class="codepoint">U+0022</code>), and
+          <code>\</code> (backslash, code point <code class="codepoint">U+005C</code>)
+          MUST be encoded using <code><a data-cite="N-QUADS#grammar-production-ECHAR">ECHAR</a></code>.</li>
+        <li>Characters in the range from <code class="codepoint">U+0000</code> to <code class="codepoint">U+001F</code>,
+          <code>DEL</code> (delete, code point <code class="codepoint">U+007F</code>),
+          and characters not matching the <a data-cite="XML11#charsets">Char</a> production from [[XML11]]
+          that are not represented using <code><a data-cite="N-QUADS#grammar-production-ECHAR">ECHAR</a></code>
+          MUST be represented by <code><a data-cite="N-QUADS#grammar-production-UCHAR">UCHAR</a></code>
+          using the shortest possible encoding.</li>
```

Comment:
```suggestion
        <li>Characters in the range from <code class="codepoint">U+0000</code> to <code class="codepoint">U+0007</code>,
          <code>VT</code> (vertical tab, code point <code class="codepoint">U+000B</code>),
          characters in the range from <code class="codepoint">U+000E</code> to <code class="codepoint">U+001F</code>,
          <code>DEL</code> (delete, code point <code class="codepoint">U+007F</code>),
          and characters not matching the <a data-cite="XML11#charsets">Char</a> production from [[XML11]]
          MUST be represented by <code><a data-cite="N-QUADS#grammar-production-UCHAR">UCHAR</a></code>
          using a lowercase <code>\u</code> with 4 <code><a data-cite="N-QUADS#grammar-production-HEX">HEX</a></code>es.</li>
```
I've enumerated the targets of the UCHAR encoding more explicitly, and
considering that only 4-HEX UCHAR encoding appears here, I've narrowed down the expression to be specific to that.
(It's my understanding that characters not matching the XML Char production should correspond to `U+0000`, `U+D800`-`U+DFFF`, `U+FFFE`, and `U+FFFF`, all of which can be encoded using 4 HEXes.)

Reply:
I think this is missing U+000C `VT` (form feed). I'm not sure that spelling these out adds all that much. Note that XML 1.1 [Char](https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-Char) production also includes `[#x10000-#x10FFFF]`, which are represented as `\U00010000` through `\U0010FFFF`. It allows pretty much everything but surrogates. UChar doesn't include U+0000 which is explicitly included above.

The "shortest possible encoding" is intended to infer either `\uxxxx` or `\U00xxxxxx` implicitly. The [UCHAR](https://www.w3.org/TR/n-quads/#grammar-production-UCHAR) production should be adequate documentation for what is meant here.

Reply:
Lightly tweaking the end of @yamdan's suggestion
```suggestion
        <li>Characters in the range from <code class="codepoint">U+0000</code> to <code class="codepoint">U+0007</code>,
          <code>VT</code> (vertical tab, code point <code class="codepoint">U+000B</code>),
          characters in the range from <code class="codepoint">U+000E</code> to <code class="codepoint">U+001F</code>,
          <code>DEL</code> (delete, code point <code class="codepoint">U+007F</code>),
          and characters not matching the <a data-cite="XML11#charsets">Char</a> production from [[XML11]]
          MUST be represented by <code><a data-cite="N-QUADS#grammar-production-UCHAR">UCHAR</a></code>
          using a lowercase <code>\u</code> with 4 <code><a data-cite="N-QUADS#grammar-production-HEX">HEX</a></code>es
          for characters in the range from <code class="codepoint">U+0000</code> to <code class="codepoint">U+FFFF</code>
          and using an uppercase <code>\U</code> with 8 <code><a data-cite="N-QUADS#grammar-production-HEX">HEX</a></code>es
          for characters in the range from <code class="codepoint">U+00010000</code> to <code class="codepoint">U+FFFFFFFF</code>.</li>

```

Reply:
Thanks! However, I still believe that we don't need to introduce "\U with 8 HEXes" in this context.

If I recall correctly, our goal is to convey that "characters not matching the Char production from XML11 MUST be represented by UCHAR."
Given that the "[Char](https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-Char) production from XML11" is defined as:

```
Char	   ::=   	[#x1-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]
```

"Characters not matching the Char production" should equal to `#x0 | [#xD800-#xDFFF] | [#xFFFE-#xFFFF]`, all of which can be represented within 4 HEXes.
We don't need to consider values like `[#x110000-...]` as the Unicode set is defined from `[#x0-#x10FFFF]`.

I'm not a Unicode expert, so please correct me if I'm mistaken in any way.

Reply:
My thought was, in part, future proofing. Also, I think it's problematic to explicitly talk about the lowercase `\u` range and omit any mention of the uppercase `\U` range.

_If_ the largest Unicode code point will _always_ be lower than `#xFFFFFFFF`, then sure, the high point of what I wrote could be lowered to whatever that highest point will be ... which you say is `#x10FFFF`, so --
```
using an uppercase <code>\U</code> with 8 
<code><a data-cite="N-QUADS#grammar-production-HEX">HEX</a></code>es for 
characters in the range from <code class="codepoint">U+00010000</code> to 
<code class="codepoint">U+0010FFFF</code>.</li>
```
-- perhaps trimming from 8 to 6 `HEX`es, if omitting insignificant `0`s is always legal.

But I've been awake for far too long, so I may be mis-thinking something. :-)

Reply:
Yes, I was too hasty about the high-range of Unicode (danger of late-night replies). There would not be a case for using the `\U00XXXXXX` form given the requirement for using the native representation.

> I believe U+000C should be encoded as ECHAR rather than UCHAR.

The set of characters represented by ECHAR `"\" [tbnrf"']` is long established, and I don't think there's much of a desire to add to it at this point.

Regarding future-proofing, the specs reference a specific version of Unicode, which explicitly defines the different [character blocks](https://en.wikipedia.org/wiki/Unicode_block#List_of_blocks). Anything outside of this range would not be valid from XML 1.1, which is our basis, so adding provision for an encoding that can't be used isn't too useful.